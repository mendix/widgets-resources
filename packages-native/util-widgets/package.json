{
    "name": "@native-mobile-resources/util-widgets",
    "version": "0.0.1",
    "private": true,
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "scripts": {
        "start": "tsc --watch",
        "prebuild": "rm -rf ./dist",
        "build": "tsc",
        "lint": "eslint --config ../../.eslintrc.js --ext .jsx,.js,.ts,.tsx src/",
        "prepare": "npm run build"
    },
    "dependencies": {
        "eslint": "^6.1.0",
        "typescript": "^3.5.3"
    }
}
