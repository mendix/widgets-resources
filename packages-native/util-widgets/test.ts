import { ActionValue, DynamicValue, EditableValue, ValueStatus } from "mendix";

export function dynamicValue<T>(value?: T): DynamicValue<T> {
    return value == null ? { status: ValueStatus.Loading, value: undefined } : { status: ValueStatus.Available, value };
}

export function actionValue(canExecute = true, isExecuting = false): ActionValue {
    return { canExecute, isExecuting, execute: jest.fn() };
}

export class EditableValueBuilder<T extends string | boolean | Date | BigJs.Big> {
    private readonly editableValue: EditableValue<T> = {
        value: undefined,
        displayValue: "",
        status: ValueStatus.Available,
        validation: undefined,
        readOnly: false,
        formatter: {
            format: jest.fn(name => `Formatted ${name}`),
            parse: jest.fn(),
            getFormatPlaceholder: jest.fn()
        },
        setValidator: jest.fn(),
        setValue: jest.fn((value: T) => this.withValue(value)),
        setTextValue: jest.fn(),
        setFormatter: jest.fn()
    };

    withValue(value?: T): EditableValueBuilder<T> {
        this.editableValue.value = value;
        this.editableValue.displayValue = this.editableValue.formatter.format(value);
        return this;
    }

    isReadOnly(): EditableValueBuilder<T> {
        this.editableValue.readOnly = true;
        return this;
    }

    isLoading(): EditableValueBuilder<T> {
        this.editableValue.status = ValueStatus.Loading;
        return this.isReadOnly();
    }

    isUnavailable(): EditableValueBuilder<T> {
        this.editableValue.status = ValueStatus.Unavailable;
        return this.isReadOnly();
    }

    withValidation(validation?: string): EditableValueBuilder<T> {
        this.editableValue.validation = validation;
        return this;
    }

    withUniverse(...values: T[]): EditableValueBuilder<T> {
        this.editableValue.universe = values;
        return this;
    }

    build(): EditableValue<T> {
        return this.editableValue;
    }
}
