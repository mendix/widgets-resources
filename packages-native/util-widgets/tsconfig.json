{
    "extends": "@mendix/pluggable-widgets-tools/configs/tsconfig.base",
    "baseUrl": "./",
    "include": ["./src"],
    "compilerOptions": {
        "target": "es2015",
        "lib": ["es2015"],
        "types": ["mendix-client", "big.js", "react-native", "jest"],
        "moduleResolution": "node",
        "module": "es2015",
        "allowSyntheticDefaultImports": true,
        "esModuleInterop": true,
        "strict": true,
        "noEmitOnError": true,
        "noFallthroughCasesInSwitch": true,
        "noImplicitAny": true,
        "noImplicitReturns": true,
        "noImplicitThis": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "forceConsistentCasingInFileNames": true,
        "skipLibCheck": true,
        "jsx": "react",
        "jsxFactory": "createElement",
        "baseUrl": "./",
        "outDir": "./dist",
        "declaration": true,
        "sourceMap": false
    }
}
