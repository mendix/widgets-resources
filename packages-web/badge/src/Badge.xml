<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.custom.badge.Badge" needsEntityContext="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../xsd/widget.xsd">
    <name>Badge</name>
    <description>Badges are numerical indicators of how many items are associated with a link</description>
    <icon>
        iVBORw0KGgoAAAANSUhEUgAAABAAAAAMCAYAAABr5z2BAAAAAXNSR0IArs4c6QAAASxJREFUKBWVUjFOxDAQ3LU3l0C4q2kRICFqmuMhUPMhPkBBeeILlAiJ5iRoQEh0FDzgIAh0WZvZGCMuBQpjWbtZz0x24zBNT9cOpvvn4436OGhLQ+C80OtbczG/eTgRqeLh1s7m0ePLO0V2FLEYy6KBsS2zmMGBaQ+au9v7M4nt51iV2YtRftMSnVGKcLDY4duthca0QhRixOF6iRSIYDPYFv+C6xxDFA+WBqJqZNlweO/IFN1r2fGPgXW6MnOvkMdx0Gg2ELgVPg3ZHyGPY3EV6VkUPiUG8UWfkOi5m/4X0aWdaxqhLD1JdwuomU9WGSfn2cE6wRytSwUhVSqKEdWV3f0w2DuaD7ChFX2+nl9dbs+KerIb22F/IovQslk8mdbMbE+wa+x/NQH+4gsBPW0QwdZQLAAAAABJRU5ErkJggg==
    </icon>
    <properties>
        <property key="label" type="translatableString" required="false">
            <caption>Label</caption>
            <category>Display</category>
            <description>The badge label</description>
            <translations>
                <translation lang="en_US"></translation>
                <translation lang="nl_NL"></translation>
            </translations>
        </property>
        <property key="badgeValue" type="translatableString" required="false">
            <caption>Value</caption>
            <category>Display</category>
            <description>The badge value</description>
            <translations>
                <translation lang="en_US"></translation>
                <translation lang="nl_NL"></translation>
            </translations>
        </property>
        <property key="bootstrapStyle" type="enumeration" defaultValue="default">
            <caption>Bootstrap style</caption>
            <category>Display</category>
            <description>The appearance of the badge based on Bootstrap classes</description>
            <enumerationValues>
                <enumerationValue key="default">Default</enumerationValue>
                <enumerationValue key="primary">Primary</enumerationValue>
                <enumerationValue key="success">Success</enumerationValue>
                <enumerationValue key="info">Info</enumerationValue>
                <enumerationValue key="warning">Warning</enumerationValue>
                <enumerationValue key="danger">Danger</enumerationValue>
            </enumerationValues>
        </property>
        <property key="badgeType" type="enumeration" required="true" defaultValue="badge">  
            <caption>Type</caption>
            <category>Display</category>
            <description>Render it as either a badge or a color label</description>
            <enumerationValues>
                <enumerationValue key="badge">Badge</enumerationValue>
                <enumerationValue key="label">Label</enumerationValue>
                </enumerationValues>
        </property>
        <property key="labelAttribute" type="attribute" required="false">
            <caption>Label attribute</caption>
            <category>Data source</category>
            <description>The attribute that contains the badge label</description>
            <attributeTypes>
                <attributeType name="Enum" />
                <attributeType name="String" />
            </attributeTypes>
        </property>
        <property key="valueAttribute" type="attribute" required="false">
            <caption>Value attribute</caption>
            <category>Data source</category>
            <description>The attribute that contains the badge value</description>
            <attributeTypes>
                <attributeType name="String" />
                <attributeType name="Integer" />
                <attributeType name="Decimal" />
                <attributeType name="Long" />
            </attributeTypes>
        </property>
        <property key="bootstrapStyleAttribute" type="attribute" required="false">
            <caption>Bootstrap style attribute</caption>
            <category>Data source</category>
            <description>The attribute value should be: default, primary, info, warning, danger, success.</description>
            <attributeTypes>
                <attributeType name="Enum" />
                <attributeType name="String" />
            </attributeTypes>
        </property>
        <property key="onClickEvent" type="enumeration" defaultValue="doNothing">
            <caption>On click</caption>
            <category>Events</category>
            <description/>
            <enumerationValues>
                <enumerationValue key="doNothing">Do nothing</enumerationValue>
                <enumerationValue key="showPage">Show a page</enumerationValue>
                <enumerationValue key="callMicroflow">Call a microflow</enumerationValue>
            </enumerationValues>
        </property>
        <property key="microflow" type="microflow" required="false">
            <caption>Microflow</caption>
            <category>Events</category>
            <description>The microflow that will be executed on click</description>
            <returnType type="Void" />
        </property>
        <property key="page" type="form" required="false">
            <caption>Page</caption>
            <category>Events</category>
            <description>The page that will be opened on click</description>
        </property>
    </properties>
</widget>
