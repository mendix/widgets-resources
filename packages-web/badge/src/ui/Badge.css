.widget-badge-clickable {
    cursor: pointer;
}

.widget-badge {
    display: inline-block;
}

.widget-badge.badge:empty {
    display: initial;
    /* Fix padding to stay round */
    padding: 3px 10px;
}

.widget-badge.label:empty {
    display: initial;
    /* Fix padding to stay square */
    padding: .2em .8em .3em;
}

.widget-badge.badge {
    min-width: 18px;
    min-height: 18px;
}
