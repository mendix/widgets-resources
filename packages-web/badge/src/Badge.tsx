import { hot } from "react-hot-loader/root";
import { ReactNode, useCallback, createElement } from "react";
import { BadgeContainerProps } from "../typings/BadgeProps";
import { Badge as DisplayBadge } from "./components/Badge";
import { ValueStatus } from "mendix";
import { executeAction } from "@widgets-resources/piw-utils";

const Badge = (props: BadgeContainerProps): ReactNode => {
    const onClick = useCallback(() => {
        executeAction(props.onClick);
    }, [props.onClick]);

    return (
        <DisplayBadge
            type={props.type}
            defaultValue={
                props.defaultValue && props.defaultValue.status === ValueStatus.Available
                    ? props.defaultValue.value
                    : undefined
            }
            value={
                props.valueAttribute && props.valueAttribute.status === ValueStatus.Available
                    ? String(props.valueAttribute.value)
                    : undefined
            }
            bootstrapStyle={props.bootstrapStyle}
            clickable={props.onClick.canExecute}
            onClick={onClick}
            className={props.class}
            style={props.style}
        />
    );
};

export default hot(Badge);
