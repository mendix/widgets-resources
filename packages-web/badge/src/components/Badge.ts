import { SFC, createElement } from "react";
import * as classNames from "classnames";

import "../ui/Badge.css";

export interface BadgeProps {
    badgeType: "badge" | "label";
    defaultValue?: string;
    className?: string;
    style?: object;
    value?: string;
    bootstrapStyle?: BootstrapStyle;
    clickable?: boolean;
    onClickAction?: () => void;
    getRef?: (node: HTMLElement) => void;
}

export type BootstrapStyle = "default" | "info" | "inverse" | "primary" | "danger" | "success" | "warning";

export const Badge: SFC<BadgeProps> = (props) => createElement("span",
    {
        className: classNames("widget-badge", props.badgeType, {
            [`label-${props.bootstrapStyle}`]: !!props.bootstrapStyle,
            "widget-badge-clickable": props.clickable
        }),
        onClick: props.onClickAction,
        ref: props.getRef,
        style: props.style
    }, props.value || props.defaultValue);
