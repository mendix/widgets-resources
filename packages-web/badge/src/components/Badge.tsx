import { createElement, ReactElement } from "react";
import classNames from "classnames";

import "../ui/Badge.css";

export interface BadgeProps {
    type: "badge" | "label";
    defaultValue?: string;
    className?: string;
    style?: object;
    value?: string;
    bootstrapStyle?: BootstrapStyle;
    clickable?: boolean;
    onClick?: () => void;
    getRef?: (node: HTMLElement) => void;
}

export type BootstrapStyle = "default" | "info" | "inverse" | "primary" | "danger" | "success" | "warning";

export const Badge = (props: BadgeProps): ReactElement =>
    createElement(
        "span",
        {
            className: classNames("widget-badge", props.type, props.className, {
                [`label-${props.bootstrapStyle}`]: !!props.bootstrapStyle,
                "widget-badge-clickable": props.clickable
            }),
            onClick: props.onClick,
            ref: props.getRef,
            style: props.style
        },
        props.value || props.defaultValue
    );
