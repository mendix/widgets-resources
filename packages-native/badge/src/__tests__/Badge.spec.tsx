import { actionValue, dynamicValue } from "@native-mobile-resources/util-widgets";
import { createElement } from "react";
import { Platform, Text } from "react-native";
import { fireEvent, render } from "react-native-testing-library";

import { Badge, Props } from "../Badge";

jest.mock("Platform", () => ({
    OS: "ios",
    select: jest.fn(dict => dict.ios)
}));

const defaultProps: Props = {
    name: "badge-test",
    style: [],
    caption: dynamicValue<string>(true)
};

describe("Badge", () => {
    it("renders with default styles", () => {
        const component = render(<Badge {...defaultProps} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders with custom styles", () => {
        const style = [
            {
                container: { backgroundColor: "red", borderColor: "green" },
                badge: { backgroundColor: "red" },
                caption: { color: "red" }
            },
            {
                container: { backgroundColor: "green" },
                badge: { backgroundColor: "green" },
                caption: { color: "green" }
            }
        ];
        const component = render(<Badge {...defaultProps} style={style} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    describe("renders a caption", () => {
        const component = render(<Badge {...defaultProps} caption={dynamicValue(false, "Caption")} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    describe("with on click action", () => {
        describe("iOS", () => {
            it("renders touchables", () => {
                const onClickAction = actionValue();
                const component = render(<Badge {...defaultProps} onClick={onClickAction} />);
                expect(component.toJSON()).toMatchSnapshot();
            });

            it("executes the action when pressed", () => {
                const onClickAction = actionValue();
                const component = render(<Badge {...defaultProps} onClick={onClickAction} />);
                fireEvent.press(component.getByType(Text));
                expect(onClickAction.execute).toHaveBeenCalledTimes(1);
            });
        });

        describe("Android", () => {
            beforeEach(() => {
                Platform.OS = "android";
                Platform.select = jest.fn(dict => dict.android!);
            });

            it("renders touchables", () => {
                const onClickAction = actionValue();
                const component = render(<Badge {...defaultProps} onClick={onClickAction} />);
                expect(component.toJSON()).toMatchSnapshot();
            });

            it("executes the action when pressed", () => {
                const onClickAction = actionValue();
                const component = render(<Badge {...defaultProps} onClick={onClickAction} />);
                fireEvent.press(component.getByType(Text));
                expect(onClickAction.execute).toHaveBeenCalledTimes(1);
            });
        });
    });
});
