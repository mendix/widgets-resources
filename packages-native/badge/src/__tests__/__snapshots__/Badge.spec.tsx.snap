// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<View
  style={
    Object {
      "alignSelf": "baseline",
      "backgroundColor": "#ccc",
      "borderRadius": 30,
      "justifyContent": "center",
      "paddingHorizontal": 10,
      "paddingVertical": 5,
    }
  }
  testID="badge-test"
>
  <Text
    style={
      Object {
        "color": "#444",
        "fontSize": 14,
        "fontWeight": "bold",
      }
    }
    testID="badge-test$caption"
  >
    Caption
  </Text>
</View>
`;

exports[`Badge renders with custom styles 1`] = `
<View
  style={
    Object {
      "alignSelf": "baseline",
      "backgroundColor": "green",
      "borderColor": "green",
      "borderRadius": 30,
      "justifyContent": "center",
      "paddingHorizontal": 10,
      "paddingVertical": 5,
    }
  }
  testID="badge-test"
>
  <Text
    style={
      Object {
        "color": "green",
        "fontSize": 14,
        "fontWeight": "bold",
      }
    }
    testID="badge-test$caption"
  >
    
  </Text>
</View>
`;

exports[`Badge renders with default styles 1`] = `
<View
  style={
    Object {
      "alignSelf": "baseline",
      "backgroundColor": "#ccc",
      "borderRadius": 30,
      "justifyContent": "center",
      "paddingHorizontal": 10,
      "paddingVertical": 5,
    }
  }
  testID="badge-test"
>
  <Text
    style={
      Object {
        "color": "#444",
        "fontSize": 14,
        "fontWeight": "bold",
      }
    }
    testID="badge-test$caption"
  >
    
  </Text>
</View>
`;

exports[`Badge with on click action Android renders touchables 1`] = `
<View
  style={
    Object {
      "alignSelf": "baseline",
      "backgroundColor": "#ccc",
      "borderRadius": 30,
      "justifyContent": "center",
      "paddingHorizontal": 10,
      "paddingVertical": 5,
    }
  }
  testID="badge-test"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffbcbc",
          "borderColor": "red",
          "borderWidth": 1,
          "height": 100,
          "justifyContent": "center",
          "margin": 10,
          "width": 300,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#333333",
          "margin": 20,
        }
      }
    >
      TouchableNativeFeedback is not supported on this platform!
    </Text>
  </View>
</View>
`;

exports[`Badge with on click action iOS renders touchables 1`] = `
<View
  style={
    Object {
      "alignSelf": "baseline",
      "backgroundColor": "#ccc",
      "borderRadius": 30,
      "justifyContent": "center",
      "paddingHorizontal": 10,
      "paddingVertical": 5,
    }
  }
  testID="badge-test"
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#444",
          "fontSize": 14,
          "fontWeight": "bold",
        }
      }
      testID="badge-test$caption"
    >
      
    </Text>
  </View>
</View>
`;
